Move JSON formatting code to Runtime Module

carlpeto opened the pull request:
Refactoring the JSON writer in `swift-backtrace` to allow its reuse. This will be used by the new swift-symbolicate tool.

Currently this is in draft for feedback and is not yet complete. The shared JSON formatting code has been moved into the Runtime module.

This now also simplifies the flow a bit, at least for JSON. The crash log is all captured in the CrashLog instance and output to JSON from there, meaning the exact same code will be shared between swift-backtrace and swift-symbolicate for JSON output, rather than workarounds.

carlpeto committed:
refactor JSON.swift from swift-backtrace to be useable by swift-symbolicate

carlpeto committed:
complete refactoring of JSON formatting code out from swift-backtrace specific code (prior to move)

carlpeto committed:
move JSON formatter into Runtime module

al45tair commented on stdlib/public/RuntimeModule/BacktraceJSONFormatter.swift:
```diff
@@ -0,0 +1,322 @@
+//  Provides functionality to format JSON backtraces.
+//
+//===----------------------------------------------------------------------===//
+import Swift
```
Nit: Let's have a blank line above this:

```suggestion

import Swift
```

al45tair commented on stdlib/public/RuntimeModule/BacktraceJSONFormatter.swift:
```diff
@@ -0,0 +1,322 @@
+
+  static func writeThreads(
+    mentionedImages: inout Set<Int>,
+    capturedBytes: inout [UInt64:Array<UInt8>]) {
```
Why are these `inout`? We don't modify them, do we?

al45tair reviewed (commented):
I think we should make `BacktraceJSONFormatter` a little bit more like `BacktraceFormatter`, API-wise; I don't think it should be a `protocol`, with users of the type expected to implement methods from that protocol — rather, I think a lot of those functions should really be handled via a `BacktraceJSONFormatterOptions` struct.

Also, this PR needs a proper title :-) But you knew that already.

carlpeto committed:
pr feedback

Changed the way swift-backtrace works for JSON, so that it captures
into a CrashLog, then writes to JSON from the CrashLog. Separating out
the two sets of logic so that writing the CrashLog to JSON becomes
simpler and more modular.

carlpeto committed:
pr feedback and unit test fixes

carlpeto commented:
@swift-ci please smoke test

carlpeto commented:
@swift-ci please smoke test

carlpeto committed:
make sure crashed is explicitly set if absent, and that symbolication options can be passed in
